nodes/src/nodes/dataset_cobalt/dataset_loader.py

```
@@ -0,0 +1,294 @@
+from rocketlib import debug, warning
+
+
+def _validate_path(path: str) -> str:
+    """Validate and return the canonical path, raising ValueError if outside cwd.
+
+    Resolves symlinks via os.path.realpath() and verifies the result lives
+    under the current working directory. Also rejects raw '..' components
+    before normalisation (they would be resolved away by realpath).
+    """
+    if '..' in path.replace('\\', '/').split('/'):
+        raise ValueError(f'Path traversal detected in file path: {path}')
```

Ruff (0.15.10)

[warning] Avoid specifying long messages outside the exception class

(TRY003)